### PR TITLE
[SPARK-37391][SQL]JdbcConnectionProvider tells if it modifies security context

### PR DESCRIPTION
**Pull request by tdg5**
Augments the JdbcConnectionProvider API such that a provider can indicate that it will need to modify the global security configuration when establishing a connection, and as such, if access to the global security configuration should be synchronized to prevent races.

### What changes were proposed in this pull request?
As suggested by @gaborgsomogyi [here](https://github.com/apache/spark/pull/29024/files#r755788709), augments the `JdbcConnectionProvider` API to include a `modifiesSecurityContext` method that can be used by `ConnectionProvider` to determine when `SecurityConfigurationLock.synchronized` is required to avoid race conditions when establishing a JDBC connection.

### Why are the changes needed?
Provides a path forward for working around a significant bottleneck introduced by synchronizing `SecurityConfigurationLock` every time a connection is established. The synchronization isn't always needed and it should be at the discretion of the `JdbcConnectionProvider` to determine when locking is necessary. See [SPARK-37391](https://issues.apache.org/jira/browse/SPARK-37391) or [this thread](https://github.com/apache/spark/pull/29024/files#r754441783).

### Does this PR introduce _any_ user-facing change?
Any existing implementations of `JdbcConnectionProvider` will need to add a definition of `modifiesSecurityContext`. I'm also open to adding a default implementation, but it seemed to me that requiring an explicit implementation of the method was preferable.

A drop-in implementation that would continue the existing behavior is:
```scala
override def modifiesSecurityContext(
  driver: Driver,
  options: Map[String, String]
): Boolean = true
```

### How was this patch tested?
Unit tests, but I also plan to run a real workflow once I get the initial thumbs up on this implementation.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcConnectionProvider.scala
Comment (tdg5) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcConnectionProvider.scala:76]: I'm open to other naming ideas.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/BasicConnectionProvider.scala
Comment (tdg5) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/BasicConnectionProvider.scala:57]: This could also just return `false`, but I'm kind of confused about what's going on here. [This doc](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/jdbc/README.md#:~:text=Built%2Din%20CPs%20added%20which%20support%20kerberos%20authentication%20using%20keytab%20and%20principal) suggests that built-in CPs support kerberos auth, but if I understand correctly that should require a `keytab` and ` principal`. I'm inclined to believe that this `JdbcConnectionProvider` doesn't actually support kerberos auth, but maybe I'm misunderstanding.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProviderSuite.scala
Comment (tdg5) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProviderSuite.scala:132]: Open to better ideas on testing that the lock was acquired or not. Could change it to an actual lock, but preferred to avoid the downstream repercussions.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcConnectionProvider.scala
Comment (tdg5) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcConnectionProvider.scala:76]: Also open to a default implementation, but seemed better to force implementers to explicitly implement.

**Comment by tdg5** [comment]
@gaborgsomogyi

**Comment by tdg5** [comment]
This is not a work in progress as far as I am concerned/aware, but I marked it as a WIP until @gaborgsomogyi can give it an initial thumbs up.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/BasicConnectionProvider.scala
Comment (tdg5) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/BasicConnectionProvider.scala:57]: Oops, just realized while driving that I need to invert this check. Well fix when I'm back at the keyboard.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/BasicConnectionProvider.scala
Comment (tdg5) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/BasicConnectionProvider.scala:57]: This could also just return `false`, but I'm kind of confused about what's going on here. [This doc](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/jdbc/README.md#:~:text=Built%2Din%20CPs%20added%20which%20support%20kerberos%20authentication%20using%20keytab%20and%20principal) suggests that built-in CPs support kerberos auth, but if I understand correctly that should require a `keytab` and `principal`. I'm inclined to believe that this `JdbcConnectionProvider` doesn't actually support kerberos auth, but maybe I'm misunderstanding.

**Comment by HyukjinKwon** [comment]
ok to test

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProviderSuite.scala
Comment (LuciferYang) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProviderSuite.scala:72]: Should I write the default method to `JdbcConnectionProvider` (return false) and override it only where necessary

Comment (tdg5) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProviderSuite.scala:72]: I'm definitely open to this, but defer to the community on whether that is appropriate or not.

It's worth noting that a default of `false` seems like the right answer to me, but Spark versions 3.1+ implicitly have a default of `true`. 

I think a default of `false` is sensible, but it would change what is currently the default behavior.

**Comment by tdg5** [comment]
Hello friends! Anything I can do to help move this PR forward? Presently my ability to upgrade past spark 3.0.x is blocked by the change in functionality that this PR aims to improve.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcConnectionProvider.scala
Comment (gaborgsomogyi) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcConnectionProvider.scala:76]: Maybe we can call it `modifiesSecurityContext`.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProviderSuite.scala
Comment (gaborgsomogyi) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProviderSuite.scala:72]: Bad idea to add default implementation(at least within the trait). Such case a user may think it's not important part of the provider and will work with the default. This is not true for all the secure providers.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/BasicConnectionProvider.scala
Comment (gaborgsomogyi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/BasicConnectionProvider.scala:57]: `BasicConnectionProvider` is the default unsecure connection provider so return false is good.

**Review by gaborgsomogyi — COMMENTED** [review]
I think the direction looks good but I've some suggestions.

**Comment by tdg5** [comment]
@gaborgsomogyi I've updated the PR to include your suggestions (really just a rename and a comment on why `BasicConnectionProvider` can always return `false`). Have another look when you have a chance, please.

**Comment by gaborgsomogyi** [comment]
Have you exexuted the integration tests?

**Comment by gaborgsomogyi** [comment]
+ plz fix mima cheks...

**Comment by tdg5** [comment]
Mima check has been fixed (assuming adding an exclude is the appropriate solution). Any idea if the kubernetes failures are something to worry about? I can't tell from the output if that's related to my change. Seems far removed from what I can tell, but it can be hard to tell. I've rebased from master, so perhaps that will help also.

**Comment by tdg5** [comment]
I think the k8s integration tests are generally borked. If you happen to know that's not true, let me know and I can investigate further.
image

**Comment by tdg5** [comment]
Any trick to re-run tests? The spark test that failed in Jenkins passes for me locally. Maybe a transient failure. No rush, I'll force push something if I don't hear back.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
@tdg5 mind taking a look at https://github.com/apache/spark/pull/34745/checks?check_run_id=4554725898 and enable Github Actions in your forked repository? Apache Spark repo uses the GitHub Actions resources of forked repository to run the build.

**Comment by tdg5** [comment]
SparkR test failure doesn't seem to relate to my commit... Rebased from master to force another build.

SparkR test failure seems to have been an ephemeral thing?
```
* installing the package to build vignettes
* creating vignettes ...Warning in file(con, "r") :
  cannot open file '/tmp/RtmpTxszZ3/xshell185ef769eca3b6': No such file or directory
Error in file(con, "r") : cannot open the connection
Error in setwd(owd) : cannot change working directory
Execution halted
sh: 0: getcwd() failed: No such file or directory
```

**Comment by gaborgsomogyi** [comment]
Some of the tests are flaky but we need a green build to merge anything so retest is the way.

Other than that plz update the description to reflect the latest code.

**Comment by tdg5** [comment]
Description updated and another round of tests is underway.

**Comment by tdg5** [comment]
Wahoo! All tests passed on the last build. Well, not the k8s integration tests, but I don't get the impression those are blocking. 🎉

**Comment by gaborgsomogyi** [comment]
There are several manual integration tests which cover this feature, have you executed them manually?

**Comment by tdg5** [comment]
> There are several manual integration tests which cover this feature, have you executed them manually?

No, I was unaware of those. Any guidance available for running those? I will look into it next time I'm at a keyboard.

**Comment by tdg5** [comment]
Something like this I'm guessing?
```
ENABLE_DOCKER_INTEGRATION_TESTS=1 \
POSTGRES_DOCKER_IMAGE_NAME=postgres:14.0 \
MYSQL_DOCKER_IMAGE_NAME=mysql:5.7.36 \
MSSQLSERVER_DOCKER_IMAGE_NAME=2019-CU13-ubuntu-20.04 \
MARIADB_DOCKER_IMAGE_NAME=mariadb:10.5.12 \
DB2_DOCKER_IMAGE_NAME=ibmcom/db2:11.5.6.0a \
ORACLE_DOCKER_IMAGE_NAME=gvenzl/oracle-xe:latest \
./build/sbt -mem 4096 -Pdocker-integration-tests "testOnly *PostgresKrbIntegrationSuite org.apache.spark.sql.jdbc.PostgresIntegrationSuite org.apache.spark.sql.jdbc.MySQLIntegrationSuite org.apache.spark.sql.jdbc.MsSqlServerIntegrationSuite org.apache.spark.sql.jdbc.MariaDBKrbIntegrationSuite *DB2KrbIntegrationSuite org.apache.spark.sql.jdbc.DB2IntegrationSuite org.apache.spark.sql.jdbc.OracleIntegrationSuite"
```

If so, I'm trying that now. Seems like a lib and some referenced docker images are out of date. I'm going to open another PR with those updates so I can play unit test roulette in a context I deem less time sensitive than this one.

**Comment by tdg5** [comment]
If I identified the correct integration tests, I can confirm they all pass on this branch when run manually. I'm out of time for the moment, so will have to wait to open the PR with the changes needed to make the docker integration tests work, but here's the diff so you can validate I didn't do anything crazy: https://github.com/tdg5/spark/commit/1d6fff5e313eb35541c2c04054b9cf9291c67962

**Comment by HyukjinKwon** [comment]
@tdg5, thanks i think it's getting closer to be merged. but just to be sure, would you mind enabling GitHub Actions in your forked repository? See also https://github.com/apache/spark/pull/34745/checks?check_run_id=4569440425. Jenkins will be retired soon, and we're currently running GitHub Actions for tests (that actually covers more correctly).

I haven't followed the changes and issue here to be honest so I will leave this to @gaborgsomogyi.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcConnectionProvider.scala
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcConnectionProvider.scala:74]: Let's also mention `@since 3.3.0`. Otherwise, looks pretty good.

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcConnectionProvider.scala:74]: I don't mind backporting it with, for example, `@since 3.2.1` or `@since 3.1.3` if you think this is a serious issue. I will leave this to you too @gaborgsomogyi.

Comment (gaborgsomogyi) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcConnectionProvider.scala:74]: I think it worth a backport since it's a performance regression issue.

**Comment by gaborgsomogyi** [comment]
The changeset looks good, except the `@since...` stuff. @tdg5 just to double check the last and most important thing do you have some performance numbers how much is it improved in unsecure case?

**Comment by tdg5** [comment]
I added the missing `@since` with a value of `@since 3.1.3`. I also corrected the MIMA exception that I had added because I put it in the 3.1->3.2 bucket instead of the 3.2->3.3 bucket. Not sure if the `@since` changes that, but the MIMA test failed w/o the exception w/ the `@since` in place, so I imagine it is still required.

Not sure why the GitHub actions aren't working on my fork; they are enabled. I tweaked a couple of other things. Hopefully that helps.

Since I have to play unit test roulette due to the addition of `@since`, I added the library changes to make the docker integration tests work to this PR also. Let me know if that's a problem or if you have any thoughts on the changes.

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala
Comment (tdg5) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala:30]: As far as I could tell the referenced image no longer exists in docker hub.

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala
Comment (tdg5) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala:32]: As far as I could tell the referenced image needs to be pulled from mcr.microsoft.com and the previously referenced image doesn't seem to exist anymore. https://hub.docker.com/_/microsoft-mssql-server

File: project/SparkBuild.scala
Comment (tdg5) [inline project/SparkBuild.scala:603]: The previously referenced version no longer exists in jfrog, so I updated to the closest similar version that was available. https://camunda.jfrog.io/ui/native/public/com/oracle/ojdbc6

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala
Comment (tdg5) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala:32]: Also, this example was missing the repository name, so I added that

**Comment by tdg5** [comment]
@HyukjinKwon , as far as I can tell, it seems I succeeded in properly enabling GitHub actions for my fork 👍🏻 If you still see otherwise, please let me know.

@gaborgsomogyi , can you offer some thoughts on what kind of performance stats are you looking for? Since this is a threading/locking issue, the speed up of being able to open more connections simultaneously can be kind of unlimited depending on how many threads are allocated to doing that kind of work. I don't mind trying to come up with some numbers, but I'm not sure what makes sense to use as a baseline.

**Comment by tdg5** [comment]
Sorry, I think this is a dumb question, but I can't seem to find the answer with google. Is there a nightly snapshot build for 3.3 I can download somewhere?

**Comment by HyukjinKwon** [comment]
> Sorry, I think this is a dumb question, but I can't seem to find the answer with google. Is there a nightly snapshot build for 3.3 I can download somewhere?

It's not a dumb question :-) - I believe most of people don't know. Here you go: https://repository.apache.org/content/groups/snapshots/org/apache/spark

POM would have to have something like:

```xml
<repository>
  <id>apache-snapshot-repo</id>
  <name>Apache Snapshot Repository</name>
  <url>https://repository.apache.org/content/groups/snapshots</url>
  <releases>
    <enabled>false</enabled>
  </releases>
  <snapshots>
    <enabled>true</enabled>
  </snapshots>
</repository>
```

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala
Comment (HyukjinKwon) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala:30]: cc @sarutak

File: project/SparkBuild.scala
Comment (HyukjinKwon) [reply project/SparkBuild.scala:603]: cc @sarutak FYI

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala
Comment (HyukjinKwon) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala:32]: cc @sarutak FYI

**Comment by HyukjinKwon** [comment]
But I think current state of CI seems fine enough to go and merge.

File: project/SparkBuild.scala
Comment (HyukjinKwon) [reply project/SparkBuild.scala:603]: BTW, is it something we can backport up to branch-3.1? If not, let's separate these changes into a separate PR so that can be only merged to master or other branches.

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala
Comment (sarutak) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala:30]: `10.5.12` is still available.
https://hub.docker.com/layers/mariadb/library/mariadb/10.5.12/images/sha256-a5ebad8f03ea9352ccf1ec3c84d91c2112ba86bbface1b57091f37aa7899d6c9?context=explore

But it's OK to upgrade.

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala
Comment (sarutak) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala:32]: @tdg5 Thank you for fixing the example. Could you fix the example in `v2/MsSqlServerIntegrationSite` too?

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala
Comment (sarutak) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala:30]: @tdg5 If you'd like to change here, please change [this part](https://github.com/apache/spark/pull/34745/files#diff-cb12fa58307c2155fe2156d347d7a7917f1c755c34488a741dfb06653f83f364L41) too.

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala
Comment (sarutak) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala:32]: Also, please change [this part](https://github.com/apache/spark/pull/34745/files#diff-cb12fa58307c2155fe2156d347d7a7917f1c755c34488a741dfb06653f83f364L40) too.

File: project/SparkBuild.scala
Comment (sarutak) [reply project/SparkBuild.scala:603]: Hmm, I'm confused about this part. Do we need this dependency setting here?
`pom.xml` has the setting for `ojdbc8`.

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala
Comment (tdg5) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala]: I'm going to roll this back since it is not required; sorry for the confusion.

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala
Comment (tdg5) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala]: hmm, this image also seems to exist. I'm going to roll that part back, but will fix the docs in another PR.

File: project/SparkBuild.scala
Comment (tdg5) [reply project/SparkBuild.scala]: I'm going to pull this into another PR for the sake of simplifying any backporting. I will update with a link shortly.

**Comment by tdg5** [comment]
The updated changes related to running docker integration tests can now be found over here: https://github.com/apache/spark/pull/34979

**Comment by tdg5** [comment]
I should have asked this earlier; how do backports work? Is there more y'all need from me in that regard? I've checked and the code in this PR doesn't cherry pick cleanly into either `branch-3.1` or `branch-3.2`.

**Comment by HyukjinKwon** [comment]
> I should have asked this earlier; how do backports work? Is there more y'all need from me in that regard? I've checked and the code in this PR doesn't cherry pick cleanly into either `branch-3.1` or `branch-3.2`.

Yeah, it's all as you said. When committers merge, they usually manually resolve conflicts in their local if it's minor. If there's a major conflict, usually they ask the author to create another PR to `branch-3.1` or `branch-3.2`. While we can merge them as is first and you can create backporting PRs, it is often easier to split a PR to multiple PRs that can be backported and not backported.

**Comment by HyukjinKwon** [comment]
Thanks @tdg5 - i like how this patch is thoroughly tested. this is a good example to follow :-).

**Comment by tdg5** [comment]
branch-3.1 PR is here: https://github.com/apache/spark/pull/34988

**Review by HyukjinKwon — APPROVED** [review]

**Review by gaborgsomogyi — APPROVED** [review]

**Review by sarutak — APPROVED** [review]

**Comment by sarutak** [comment]
Merging to `master`. Thank you all !

**Pull request closed without merging**
